### PR TITLE
[SPARK-40809] [CONNECT] [FOLLOW] Support `alias()` in Python client

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
This extends the implementation of column aliases in Spark Connect with supporting lists of column names and providing the appropriate implementation for the Python side.

### Why are the changes needed?
Compatibility

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT in Python and Scala

### CODE REVIEW

**Comment by grundprinzip** [comment]
@amaliujia @cloud-fan @HyukjinKwon @hvanhovell

File: python/pyspark/sql/connect/column.py [hunk 1]
```diff
@@ -82,6 +82,74 @@ def to_plan(self, session: "RemoteSparkSession") -> "proto.Expression":
     def __str__(self) -> str:
         ...
 
+    def alias(self, *alias: str, **kwargs: Any) -> "Expression":
+        """
+        Returns this column aliased with a new name or names (in the case of expressions that
+        return more than one column, such as explode).
+
+        .. versionadded:: 1.3.0
```
Comment (amaliujia) [inline python/pyspark/sql/connect/column.py:90]: version 3.4.0

File: python/pyspark/sql/connect/column.py [hunk 2]
```diff
@@ -82,6 +82,74 @@ def to_plan(self, session: "RemoteSparkSession") -> "proto.Expression":
...
+
+        Other Parameters
+        ----------------
+        metadata: dict
+            a dict of information to be stored in ``metadata`` attribute of the
+            corresponding :class:`StructField <pyspark.sql.types.StructField>` (optional, keyword
+            only argument)
+
+            .. versionchanged:: 2.2.0
+               Added optional ``metadata`` argument.
```
Comment (amaliujia) [inline python/pyspark/sql/connect/column.py:105]: we don't need this: 

Connect is new API and everything will start from 3.4.0

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 3]
```diff
@@ -334,7 +334,11 @@ class SparkConnectPlanner(session: SparkSession) {
   }
 
   private def transformAlias(alias: proto.Expression.Alias): NamedExpression = {
-    Alias(transformExpression(alias.getExpr), alias.getName)()
+    if (alias.getNameCount == 1) {
+      Alias(transformExpression(alias.getExpr), alias.getName(0))()
+    } else {
+      MultiAlias(transformExpression(alias.getExpr), alias.getNameList.asScala.toSeq)
```
Comment (amaliujia) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:340]: Can you make a change in Connect DSL and add a server side test in `SparkConnectProtoSuite`? 

Should be straightforward.

File: python/pyspark/sql/connect/column.py [hunk 4]
```diff
@@ -82,6 +82,74 @@ def to_plan(self, session: "RemoteSparkSession") -> "proto.Expression":
...
+            if self._metadata:
+                raise ValueError("Creating aliases with metadata is not supported.")
+            else:
+                exp = proto.Expression()
+                exp.alias.name.append(self._alias[0])
+                exp.alias.expr.CopyFrom(self._parent.to_plan(session))
+                return exp
+        else:
+            if self._metadata:
+                raise ValueError("metadata can only be provided for a single column")
```
Comment (amaliujia) [inline python/pyspark/sql/connect/column.py:144]: Wait literally metadata is not supported yet..

Maybe just one `raise ValueError`

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 3 above)
Comment (amaliujia) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:340]: I actually expect Scala client will re-use Connect DSL. 

if you take a closer look, Connect DSL is DataFrame like internal API that offers `toProto` logic (equivalent to `plan.py` for python). 

Anything adding to the DSL will not longer help testing coverage but also help Scala client.

File: python/pyspark/sql/connect/column.py (hunk 1 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:90]: Fixed.

File: python/pyspark/sql/connect/column.py (hunk 2 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:105]: removed.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 3 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:340]: I understand the concepts of the DSL and the testing (I was there when it was written), but thanks a lot for the detailed explanation.

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:340]: I added a new test relation with a `map` type, added the type converters, test agg function and now it covers this branch as well.

File: python/pyspark/sql/connect/column.py (hunk 4 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:143]: I just pushed the implementation of metadata as well. However, now the schema does not return the metadata, but at least the input is consistent with Spark.

**Review by amaliujia — COMMENTED** [review]
LGTM

Per https://spark.apache.org/contributing.html we should follow the following when adding new tests:
```
def test_case(self):
    # SPARK-12345: a short description of the test
    ...
```

File: connector/connect/src/main/protobuf/spark/connect/expressions.proto [hunk 5]
```diff
@@ -170,6 +170,8 @@ message Expression {
 
   message Alias {
     Expression expr = 1;
-    string name = 2;
+    repeated string name = 2;
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/expressions.proto:173]: In catalyst, we have a `MultiAlias` expression for this case. Shall we match catalyst here?

Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/expressions.proto:173]: Since we're not trying to match all of the Catalyst plans but be compatible on the logical level, I find it weird to use `MultiAlias` here.

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/expressions.proto:173]: OK, then shall we fail if it's a multi alias but the metadata fied is specified?

Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/expressions.proto:173]: makes sense, let me add a server side check to the client side one

File: connector/connect/src/main/protobuf/spark/connect/expressions.proto [hunk 6]
```diff
@@ -170,6 +170,8 @@ message Expression {
...
+    // Alias metadata expressed as a JSON map.
+    optional string metadata = 3;
```
Comment (cloud-fan) [inline connector/connect/src/main/protobuf/spark/connect/expressions.proto:175]: nit: previously @amaliujia added
```
  message DataSource {
    // Required. Supported formats include: parquet, orc, text, json, parquet, csv, avro.
    string format = 1;
    // Optional. If not set, Spark will infer the schema.
    string schema = 2;
    // The key is case insensitive.
    map<string, string> options = 3;
  }
```

Shall we also have a guideline for adding string-string map field? Do we prefer proto map field or just JSON string?

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 7]
```diff
@@ -348,7 +350,16 @@ class SparkConnectPlanner(session: SparkSession) {
...
-    Alias(transformExpression(alias.getExpr), alias.getName)()
+    if (alias.getNameCount == 1) {
+      val md = if (alias.hasMetadata()) {
+        Some(Metadata.fromJson(alias.getMetadata))
+      } else {
+        None
+      }
+      Alias(transformExpression(alias.getExpr), alias.getName(0))(explicitMetadata = md)
+    } else {
+      MultiAlias(transformExpression(alias.getExpr), alias.getNameList.asScala.toSeq)
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:361]: shall we fail here if `alias.hasMetadata` is true?

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:361]: I'm adding this right now.

File: connector/connect/src/main/protobuf/spark/connect/expressions.proto (hunk 6 above)
Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/expressions.proto:175]: That's a very good question. I think in this case it's not as easy possible because it actually depends on the type and is not a plain `map<string, string>`

Comment (cloud-fan) [reply connector/connect/src/main/protobuf/spark/connect/expressions.proto:175]: ah the metadata value can be any type, then JSON LGTM.

File: connector/connect/src/main/protobuf/spark/connect/expressions.proto (hunk 5 above)
Comment (grundprinzip) [reply connector/connect/src/main/protobuf/spark/connect/expressions.proto:173]: done

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@HyukjinKwon can you review the python side?

File: python/pyspark/sql/connect/column.py [hunk 8]
```diff
@@ -82,6 +82,73 @@ def to_plan(self, session: "RemoteSparkSession") -> "proto.Expression":
...
+        self._parent = parent
+
+    def to_plan(self, session: "RemoteSparkSession") -> "proto.Expression":
+        if len(self._alias) == 1:
+            exp = proto.Expression()
+            exp.alias.name.append(self._alias[0])
+            exp.alias.expr.CopyFrom(self._parent.to_plan(session))
+
+            if self._metadata:
+                import json
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:137]: Nit but I think we can just import this on the top.

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 9]
```diff
@@ -248,6 +248,20 @@ def test_simple_datasource_read(self) -> None:
...
+        """Testing supported and unsupported alias"""
+        col0 = (
+            self.connect.range(1, 10)
+            .select(col("id").alias("name", metadata={"max": 99}))
+            .schema()
+            .names[0]
+        )
+        self.assertEqual("name", col0)
+
+        with self.assertRaises(Exception) as exc:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:261]: nit but I would catch narrower exception than `Exception`

File: python/pyspark/sql/tests/connect/test_connect_column_expressions.py [hunk 10]
```diff
@@ -134,6 +134,16 @@ def test_list_to_literal(self):
         lit_list_plan = fun.lit([fun.lit(10), fun.lit("str")]).to_plan(None)
         self.assertIsNotNone(lit_list_plan)
 
+    def test_column_alias(self) -> None:
+        # SPARK-40809 - Support for Column Aliases
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_connect_column_expressions.py:138]: nit:
```suggestion
        # SPARK-40809: Support for Column Aliases
```
per https://spark.apache.org/contributing.html

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala [hunk 11]
```diff
@@ -141,10 +149,35 @@ class SparkConnectProtoSuite extends PlanTest with SparkConnectPlanTest {
     comparePlans(connectPlan2, sparkPlan2)
   }
 
-  test("column alias") {
+  test("[SPARK-40809] column alias") {
```
Comment (HyukjinKwon) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:152]: nit but:
```suggestion
  test("SPARK-40809: column alias") {
```
per https://spark.apache.org/contributing.html

**Review by HyukjinKwon — APPROVED** [review]
LGTM some nits

File: python/pyspark/sql/tests/connect/test_connect_column_expressions.py [hunk 12]
```diff
@@ -134,6 +134,16 @@ def test_list_to_literal(self):
...
+        col0 = fun.col("a").alias("martin")
+        self.assertEqual("Alias(Column(a), (martin))", str(col0))
+
+        col0 = fun.col("a").alias("martin", metadata={"pii": True})
+        plan = col0.to_plan(self.session)
+        self.assertIsNotNone(plan)
+        self.assertEqual(plan.alias.metadata, '{"pii": true}')
```
Comment (zhengruifeng) [inline python/pyspark/sql/tests/connect/test_connect_column_expressions.py:145]: what about also adding a test for `MultiAlias` in python?

File: python/pyspark/sql/connect/dataframe.py [hunk 13]
```diff
@@ -44,7 +44,7 @@
     from pyspark.sql.connect.typing import ColumnOrString, ExpressionOrString
     from pyspark.sql.connect.client import RemoteSparkSession
 
-ColumnOrName = Union[Column, str]
+ColumnOrName = Union[Expression, str]
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/dataframe.py:47]: is this change needed?

File: python/pyspark/sql/tests/connect/test_connect_column_expressions.py (hunk 12 above)
Comment (grundprinzip) [reply python/pyspark/sql/tests/connect/test_connect_column_expressions.py:145]: See the test above.

File: python/pyspark/sql/tests/connect/test_connect_basic.py (hunk 9 above)
Comment (grundprinzip) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:261]:
```suggestion
        with self.assertRaises(grpc.RpcError) as exc:
```

File: python/pyspark/sql/connect/column.py (hunk 8 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:137]: Done.

File: python/pyspark/sql/connect/dataframe.py (hunk 13 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:47]: I think this fell out of a different PR I had for transform. The reason we need it eventually is because right now from a type perspective select can only use columns but should be able to refer to any expression.

Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py:47]:
```suggestion
ColumnOrName = Union[Column, str]
```

Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py]: tarting mypy annotations test...
annotations failed mypy checks:
python/pyspark/sql/tests/connect/test_connect_basic.py:308: error: Argument 1 to "select" of "DataFrame" has incompatible type "ColumnAlias"; expected "Union[Column, str]"  [arg-type]
python/pyspark/sql/tests/connect/test_connect_basic.py:315: error: Argument 1 to "select" of "DataFrame" has incompatible type "ColumnAlias"; expected "Union[Column, str]"  [arg-type]
F

Comment (grundprinzip) [reply python/pyspark/sql/connect/dataframe.py]: ColumnAlias is an expression and not an unresolved attribtue.

**Comment by HyukjinKwon** [comment]
Merged to master.

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 14]
```diff
@@ -301,6 +301,20 @@ def test_simple_datasource_read(self) -> None:
...
+            self.connect.range(1, 10)
+            .select(col("id").alias("name", metadata={"max": 99}))
+            .schema()
+            .names[0]
+        )
+        self.assertEqual("name", col0)
+
+        with self.assertRaises(grpc.RpcError) as exc:
+            self.connect.range(1, 10).select(col("id").alias("this", "is", "not")).collect()
+        self.assertIn("Buffer(this, is, not)", str(exc.exception))
```
Comment (LuciferYang) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:316]: @grundprinzip  The Scala 2.13 daily test has failed in recent times, which seems to be related to this assert:
```
======================================================================
FAIL [0.742s]: test_alias (pyspark.sql.tests.connect.test_connect_basic.SparkConnectTests)
Testing supported and unsupported alias
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/sql/tests/connect/test_connect_basic.py", line 337, in test_alias
    self.assertIn("Buffer(this, is, not)", str(exc.exception))
AssertionError: 'Buffer(this, is, not)' not found in '<_MultiThreadedRendezvous of RPC that terminated with:\n\tstatus = StatusCode.UNKNOWN\n\tdetails = "[INTERNAL_ERROR] Found the unresolved operator: \'Project [id#24L AS List(this, is, not)]"\n\tdebug_error_string = "UNKNOWN:Error received from peer ipv4:127.0.0.1:15002 {created_time:"2022-11-21T19:25:59.902777537+00:00", grpc_status:2, grpc_message:"[INTERNAL_ERROR] Found the unresolved operator: \\\'Project [id#24L AS List(this, is, not)]"}"\n>'
```

https://github.com/apache/spark/actions/runs/3517345801/jobs/5895043640

This seems to be related to the difference of Scala version. Can we simplify the assertion? such as `self.assertIn("(this, is, not)", str(exc.exception))`

Comment (HyukjinKwon) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:316]: I made a followup at https://github.com/apache/spark/pull/38752 :-)
